layer_store: Use CreateReadWrite() for -init layer instead of Create()

rhvgoyal opened the pull request:
init layer is read/write layer and not read only layer. Following commit
introduced new graph driver method CreateReadWrite.

ef5bfad Adding readOnly parameter to graphdriver Create method

So far only windows seem to be differentiating between above two methods.
Making this change to make sure -init layer calls right method so that
we don't have surprises in future.

Signed-off-by: Vivek Goyal vgoyal@redhat.com

rhvgoyal commented:
CC @swernli

swernli commented:
@rhvgoyal Unfortunately, I don't think this is a change we should take right now.  While this makes perfect logical sense, it has a big perf impact on Windows (creating a read/write layer is slow on Windows), and the init layer is completely unused by Windows containers (`initFunc` comes from `daemon.setupInitLayer` which is a no-op).

I think we should take this fix only after updating docker to stop creating an init layer on Windows, otherwise the perf impact is too high.

rhvgoyal commented:
@swernli If windows does not need init layer, then how about the new patch I pushed. This patch stops creation of init layer on windows.

swernli commented:
@rhvgoyal Excellent, yes this is exactly what I was hoping for.  Thanks!

LGTM.

rhvgoyal commented:
cc @dmcgowan @crosbymichael

dmcgowan commented:
Could be a smaller change to instead have a `func (daemon *Daemon) getLayerInit() func(string) error` with different implementation for windows and unix, but that seems quite nitty.

LGTM

rhvgoyal committed:
layer_store: Use CreateReadWrite() for -init layer instead of Create()

init layer is read/write layer and not read only layer. Following commit
introduced new graph driver method CreateReadWrite.

ef5bfad Adding readOnly parameter to graphdriver Create method

So far only windows seem to be differentiating between above two methods.
Making this change to make sure -init layer calls right method so that
we don't have surprises in future.

Windows does not need init layer. This patch also gets rid of creation of
init layer on windows.

Signed-off-by: Vivek Goyal <vgoyal@redhat.com>

rhvgoyal commented:
@dmcgowan I like your idea of returning function pointer. Pushed a new patch. PTAL.

swernli commented:
Even better!

dmcgowan commented:
LGTM

rhvgoyal commented:
This windowsRS1 error, I doubt it is because of this PR.

swernli commented:
@rhvgoyal A fix for that issue was actually just merged in: https://github.com/docker/docker/pull/26764
@darrenstahlmsft can confirm that this is the same, or you could rebase and push to retry CI with the latest.

darrenstahlmsft commented:
Yes, the windowsRS1 failure is fixed by that PR. I've restarted the windowsRS1 CI to pick up the fix.

thaJeztah commented on daemon/create.go:
```diff
@@ -201,7 +201,9 @@ func (daemon *Daemon) setRWLayer(container *container.Container) error {
-	rwLayer, err := daemon.layerStore.CreateRWLayer(container.ID, layerID, container.MountLabel, daemon.setupInitLayer, container.HostConfig.StorageOpt)
+
+	rwLayer, err := daemon.layerStore.CreateRWLayer(container.ID, layerID, container.MountLabel, daemon.getLayerInit(), container.HostConfig.StorageOpt)
+
```
_really_ nitty, but unrelated whitespace added?

rhvgoyal replied:
You mean, extra line added? I can get rid of it. Not a problem.

thaJeztah replied:
It's merged now, so no real need to do so (was just trying to keep the diff small - easier to check history :smile:)

thaJeztah commented on daemon/daemon_unix.go:
```diff
@@ -779,6 +779,10 @@ func initBridgeDriver(controller libnetwork.NetworkController, config *Config) e
 }
 
+func (daemon *Daemon) getLayerInit() func(string) error {
+	return daemon.setupInitLayer
```
Should probably be added to `daemon_solaris.go` as well?

rhvgoyal replied:
Hmm.., I had not noticed daemon_solaris.go. Looks like we should put it in there as well. Looks like CI does not builds for solaris? Otherwise build should have failed.

Anyway, I will update patch to take care of solaris as well.

thaJeztah replied:
It's still being worked on, so no CI yet, thanks for doing the follow up!

crosbymichael commented:
LGTM
